README.md
# Welcome to the Wartburg College Course Finder API Github
#### This project was done in accordance with the CS 460 class at Wartbrug College and supervised by Dr. Zelle
#### Main contributors are Robert Farmer and Will Goettl with assistance from Justin Schoppe and Dr. Zelle


This project aims to allow users to write scripts to manipulate and edit course data obtained form Wartburg College's course finders
at http://winnet.wartburg.edu/coursefinder/ and https://my.wartburg.edu. 
This project uses Selenium, MechanicalSoup, Pandas, and DateTimePython3 libraries to support and run our functions. 

Below are descriptions of each file, their purpose, and how they function.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

* Demo.py
  * Test file for testing of functions.

* cf_api.py
  * API for our course finder data that works with Pandas dataframe as well as contains functions to pull data from winnet. 

* chromedriver.exe
  * Required file to run silenium. See https://www.selenium.dev/documentation/en/

* course_data_4172020.sql
  * SQL export file to allow data use and manipulation in a SQL database if desiered. This flie really only exists if a user would like to
    use SQL instead of Pandas for whatever reason. I (Will) thought it would be a good idea to include just in case. 
  
* course_history.txt
  * Master course data file containing all course data to date (last update 4.17.2020)

* mywartburg_scraper.py
  * Web scraper built in silenium to scrape my.wartburg's course finder for hisorical data. Requieres a valid my.wartburg user id and pw.
   This ideally will never need to be used. This is a legacy file in case data loss or corruption occurs and a new scrape of course
   data needs to be preformed. 


Changelog.txt
"Course_ID"Course_Title"Professor_Name"Meeting_Time"Enrollment"Room"Year"Term"Credit
0"AC   121  01   "Principles of Accounting I"Pilcher, A"M W F 10:45AM-11:50AM"45/40/0"WBC  116  "2019-20"WI"1.00
1"AC   122  01   "Principles of Accounting II"Pilcher, A"MWF 12:00PM-1:05PM"45/42/0"WBC  116  "2019-20"WI"1.00
2"AC   122  02   "Principles of Accounting II"Pilcher, A"M W F 2:30PM-3:35PM"45/38/0"WBC  116  "2019-20"WI"1.00
3"AC   322  01   "Intermediate Accounting II"Crouse, J"MWF 9:00AM-10:05AM"28/27/0"WBC  116  "2019-20"WI"1.00
4"AC   323  01   "Cost Accounting"Pilcher, A" T H  9:35AM-11:15AM"30/27/0"WBC  116  "2019-20"WI"1.00
5"AC   424  01   "Income Tax Accounting II"Sommermeyer, E"M W F 10:45AM-11:50AM"40/9/0"WBC  217  "2019-20"WI"1.00
6"AC   425  01   "Auditing"Sommermeyer, E"MWF 7:45AM-8:50AM"24/14/0"WBC  217  "2019-20"WI"1.50
7"AC   425  01   "Auditing"Sommermeyer, E"  H 7:45AM-9:25AM"24/14/0"WBC  217  "2019-20"WI"1.50
8"ART  108  01   "Graphic Design"Sund, P"M W F 9:00AM-10:05AM"20/16/0"FAC  114C "2019-20"WI"1.00
9"ART  108  02   "Graphic Design"Payne, T"M W F 1:15PM-2:20PM"20/19/0"FAC  114C "2019-20"WI"1.00
10"ART  111  01   "Drawing"Fedeler, B" T H  8:30AM-11:15AM"16/17/0"FAC  114D "2019-20"WI"1.00
11"ART  201  01   "Painting"Fedeler, B" T H  1:00PM-3:35PM"20/12/0"FAC  114E "2019-20"WI"1.00
12"ART  205  01   "Communication Design"Payne, T"M W F 10:45AM-11:50AM"20/13/0"FAC  114C "2019-20"WI"1.00
13"ART  260  01   "History of Western Art"Fedeler, B"M W F 10:45AM-11:50AM"20/19/0"FAC  114J "2019-20"WI"1.00
14"ART  301  01   "Advanced Drawing and Painting"Fedeler, B" T H  1:00PM-3:35PM"18/3/0"FAC  114E "2019-20"WI"1.00
15"ART  310  01   "Advanced  Design Projects"Payne, T" T H  9:35AM-11:15AM"20/15/0"FAC  114C "2019-20"WI"1.00
16"ART  462  01   "Seminar in Art: Exhibition"Payne, T"M W   2:30PM-3:35PM"10/8/0"FAC  114J "2019-20"WI"0.50
17"ART  480  01   "K-12 Art Methods"Maynard, L"M W   6:30PM-9:00PM"18/3/0"FAC  114G "2019-20"WI"1.50
18"BA   311  01   "Comm for Business Professionals"Gonnerman, M" T    9:35AM-11:15AM"22/22/0"WBC  117  "2019-20"WI"0.50
19"BA   311  02   "Comm for Business Professionals"Gonnerman, M"   H  9:35AM-11:15AM"22/20/0"WBC  117  "2019-20"WI"0.50
20"BA   331  01   "Business Law I"Luhring, L"MWF 12:00PM-1:05PM"40/42/0"WBC  214  "2019-20"WI"1.00
21"BA   334  01   "Business Finance"Abdiah, S"MWF 9:00AM-10:05AM"36/36/0"WBC  214  "2019-20"WI"1.00
22"BA   338  01   "Portfolio Management"Sommermeyer, E"T  2:50PM-3:40PM"15/15/0"WBC  214  "2019-20"WI"0.25
23"BA   345  01   "Principles of Management"Bernard, A"MWF 9:00AM-10:05AM"32/32/0"WBC  117  "2019-20"WI"1.00
24"BA   345  02   "Principles of Management"Meyeraan, S" T H  7:45AM-9:25AM"32/27/0"WBC  117  "2019-20"WI"1.00
25"BA   346  01   "Human Resource Management"Meyeraan, S"M W F 9:00AM-10:05AM"35/20/0"WBC  215  "2019-20"WI"1.00
26"BA   353  01   "Principles of Marketing"Gonnerman, M"M W F 10:45AM-11:50AM"32/32/0"WBC  117  "2019-20"WI"1.00
27"BA   353  02   "Principles of Marketing"Folkers, K"M W F 2:30PM-3:35PM"32/28/0"WBC  215  "2019-20"WI"1.00
28"BA   361  01   "Operations Management "Black, D"M W F 7:45AM-8:50AM"25/11/0"WBC  215  "2019-20"WI"1.00
29"BA   375  01   "Online Marketing"Richardson-Burns, R" T    6:30PM-9:30PM"32/25/0"WBC  117  "2019-20"WI"1.00
30"BA   418  01   "Sustainable Economics and Finance"Quaicoe, N"M W F 10:45AM-11:50AM"30/23/0"WBC  214  "2019-20"WI"1.00
31"BA   438  01   "Organizational Behavior & Mgmt"Gonnerman, M"M W F 1:15PM-2:20PM"30/19/0"WBC  215  "2019-20"WI"1.00
32"BA   442  01   "International Marketing"Folkers, K" T H  1:00PM-2:40PM"28/13/0"WBC  217  "2019-20"WI"1.00
33"BA   454  01   "Advertising and Promotion"Folkers, K"M W F 12:00PM-1:05PM"30/30/0"WBC  215  "2019-20"WI"1.00
34"BA   460  01   "Senior Seminar in Business"Bernard, A" T H  7:45AM-9:25AM"22/22/0"WBC  215  "2019-20"WI"1.00
35"BA   460  02   "Senior Seminar in Business"Bernard, A" T H  9:35AM-11:15AM"22/23/0"WBC  215  "2019-20"WI"1.00
36"BA   481  01   "Fin Stmt Anlysis & Secur Valuation"Sommermeyer, E"M W F 1:15PM-2:20PM"40/29/0"WBC  214  "2019-20"WI"1.00
37"BI   130  01   "Drugs and Personal Health"Place, A"M F 1:15PM-2:55PM"24/25/1"SC   134  "2019-20"WI"1.00
38"BI   130  01   "Drugs and Pers Health Lab"Place, A"  W   1:15PM-2:55PM"24/25/1"SC   034  "2019-20"WI"0.00
39"BI   131  01   "Biology of Ordinary Things"Bechtel, M"M W F 9:00AM-10:05AM"24/22/0"SC   128  "2019-20"WI"1.00
40"BI   131  01   "Biology of Ordinary Things Lab"Bechtel, M" T    7:45AM-9:25AM"24/25/0"SC   128  "2019-20"WI"0.00
41"BI   131  02   "Biology of Ordinary Things"Bechtel, M"M W F 10:45AM-11:50AM"24/24/0"SC   128  "2019-20"WI"1.00
42"BI   131  02   "Biology of Ordinary Things Lab"Bechtel, M" T    9:35AM-11:15AM"24/21/0"SC   128  "2019-20"WI"0.00
43"BI   152  01   "Bio II:Phylogeny,Structure,Function"Larimer Bousquet, S"M W F 9:00AM-10:05AM"70/62/0"SC   102  "2019-20"WI"1.00
44"BI   152  01   "Bio II:Phylogeny,Structure,Func Lab"Berns, E" T    7:45AM-11:00AM"24/26/0"SC   126  "2019-20"WI"0.00
45"BI   152  02   "Bio II:Phylogeny,Structure,Function"Gerken Golay, M"M W F 12:00PM-1:05PM"50/38/0"SC   102  "2019-20"WI"1.00
46"BI   152  02   "Bio II:Phylogeny,Structure,Func Lab"Berns, E" T    1:00PM-4:15PM"24/24/1"SC   126  "2019-20"WI"0.00
47"BI   152  04   "Bio II:Phylogeny,Structure,Func Lab"Berns, E"   H  7:45AM-11:00AM"24/27/0"SC   126  "2019-20"WI"0.00
48"BI   152  05   "Bio II:Phylogeny,Structure,Func Lab"McClung, J"   H  1:00PM-4:15PM"24/23/0"SC   126  "2019-20"WI"0.00
49"BI   201  01   "Plant Biology"Gerken Golay, M"M W F 2:30PM-3:35PM"24/10/0"SC   145  "2019-20"WI"1.00
50"BI   201  01   "Plant Biology"Gerken Golay, M"   H  1:00PM-4:15PM"24/10/0"SC   145  "2019-20"WI"0.00
51"BI   204  0R   "Conserv of Natural Resources Lab"Bechtel, M"Null"1/1/0"Null"2019-20"WI"0.00
52"BI   204  2R   "Conserv of Natural Resources"McCullough, D"Null"1/1/0"Null"2019-20"WI"0.00
53"BI   211  01   "Genetics"McCullough, D"MWF 1:15PM-2:20PM"48/37/0"SC   102  "2019-20"WI"1.00

cf_api.py
#cf_api.py
#Written for CS460 01 Capstone at Wartburg College
#Instructor/Client - Dr.John Zelle
#By: Will Goettl & Rob Farmer

import mechanicalsoup
import pandas as pd
import os,time
from bs4 import BeautifulSoup as soup
from datetime import datetime

class CourseSearch:
    def __init__(self):
        '''
        Initializes Variables & StatefulBrowser for use within the CourseSearch class
        '''
        #Sets Statefulself.browser Object to winnet then it it grabs form
        self.browser = mechanicalsoup.StatefulBrowser()
        self.browser.open("http://winnet.wartburg.edu/coursefinder/")
        self.page = self.browser.select_form()
        self.option_list = self.browser.get_current_page().findAll('option')
        
        #Form Variables
        self.keyword = None
        self.dept = None
        self.term = None
        self.ee_req = None
        self.time = None
        self.is_cult_div = None
        self.ee_req = None
        self.write = False
        self.pass_fail = False
        self.instructor = None
        self.course_open = None
        
        #DropDownDictionaries
        self.course_list = []
        self.changes = []
        self.data = []
        self.Departments = {'All':'None Selected','AC   ':'Accounting', 'ART  ':'Art','BI   ':'Biology','BA   ':'Business Administration','CH   ':'Chemistry', 'CS   ':'Computer Science',
               'CCJ  ':'Criminology/Criminal Justice','EC   ':'Economics', 'ED   ':'Education', 'ES   ':'Engineering Science', 'EN   ':'English','EI   ':'English International Students',
               'ENV  ':'Environmental Science', 'EXS  ':'Exercise Science', 'FL   ':'Foreign Language','FR   ':'French','SCI  ':'General Science',
               'GER  ':'German','GM   ':'Global Multicultural','GR   ':'Greek','HE   ':'Health','HB   ':'Hebrew','HI   ':'History','IS   ':'Inquiry Studies',
               'ID   ':'Interdisciplinary','COM  ':'Journalism and Communication','LS   ':'Liberal Studies','MA   ':'Mathematics', 'MU   ':'Music',
               'NSC  ':'Neuroscience','PJ   ':'Peace and Justice Studies','PH   ':'Philiosophy','PE   ':'Physical Education','PHY  ':'Physics','PS   ':'Political Science',
               'NUR  ':'Pre-Professional Nursing Prog', 'PSY  ':'Psychology', 'PBH  ':'Public Health','RE   ':'Religion','SCH  ':'Scholars Program',
               'SW   ':'Social Work','SO   ':'Sociology','SP   ':'Spanish','TH   ':'Theatre','VE   ':'Venture Ed','WW   ':'Wartburg West','WS   ':"Womens Studies"} 
        self.Terms = ['May Term', 'Summer Session', 'Winter Term', 'Fall Term'] 
        self.Times = {'all':'Show all', 'od':'Other Evening', 'oe':'Other Evening', '7:45AM/MWF':'7:45AM MWF', '9:00AM/MWF':'9:00AM MWF', '10:45AM/MWF':'10:45AM MWF',
                    '12:00PM/MWF':'12:00PM MWF', '1:15PM/MWF':'1:15PM MWF', '2:30PM/MWF':'2:30PM MWF', '3:50PM/MWF':'3:50PM MWF', '7:45AM/THX':'7:45AM TH',
                    '7:45AM/THX':'7:45AM TH', '9:35AM/THX':'9:35AM TH','11:30AM/THX':'11:30AM TH', '1:00PM/THX':'1:00PM TH', '2:50PM/THX':'2:50PM TH', '3:50PM/THX':'3:50PM TH'}
        self.ED = {'all':'Show all EE courses', 'CP':'CP - Capstone', 'FL':'FL- Foreign Language', 'FR':'FR - Faith & Reflection',
                    'HF':'HF - Humanities/Fine Arts Interconnected','MR':'MR - Mathematical Reasoning','NS':'NS - Natural Science Interconnected',
                    'SR':'SR - Scientific Reasoning','SS':'SS - Social Science Interconnected'}
        self.CD = {'none':'Not selected','C':'C - Cultural Immersion', 'D': 'D - Diversity Across Curriculum Course','all':'Show both'}
        self.WI = {'none':'Not selected', 'WI':'Writing Intensive'}
        self.PF = {'none':'Not selected', 'PF':'Pass/D/F basis'}
        self.Instructor = {'0':'Not Selected'}
        
    def update_historic(self):
        '''
        Opens file containing historical course data
        Takes search results and indexes them into a hashtable
        It then compares the courses in winnet to the historic courses and appends
        courses that are new and overwrites courses that are pre-existing.
        '''
        historic = pd.read_csv('course_history.txt', header=None, sep='"', encoding = 'ISO-8859-1',
                               names=['Course_ID', 'Course_Title', 'Professor_Name','Meeting_Time','Enrollment','Room','Year','Term','Credit'])
        winnet = pd.DataFrame(self.data,
                              columns=['Course_ID', 'Course_Title', 'Professor_Name','Meeting_Time','Enrollment','Room','Year','Term','Credit'])
        cols = ['Course_ID','Year','Term']
        combined = pd.concat([historic,winnet])
        combined = combined.drop_duplicates(subset=cols, keep='last')
        #combined.set_index(['Course_ID','Year','Term'], inplace=True)
        combined.to_csv(r'course_history.txt',header=None,index=cols,sep='"',mode='w')
        winnet.to_csv(r'Changelog.txt',header=None,index=None,sep='"',mode='w')
        historic.to_csv(r'historic_legacy.txt',header=None,index=None,sep='"',mode='w')
        
        timestr = time.strftime("%Y-%m-%d")
        changelog = "Changelog"+timestr+".txt"
        historic_V = 'historic_legacy'+timestr+'.txt'
        #os.rename is to append the Current Date as d.dataFrame.to_csv can't take a variable str w/ Datetime
        #os.rename will not change filename if the filename already exists
        os.rename('Changelog.txt',changelog)
        os.rename('historic_legacy.txt',historic_V)
        
    def set_keyword(self,key):
        '''
        Takes a String to be passed into the Keyword search parameter
        '''
        self.page.set('ctl00$ContentPlaceHolder1$FormView1$TextBox_keyword', key)

    def get_departments(self):
        '''
        Returns current lists of departments in an array
        '''
        departs = []
        for key,value in self.Departments.items():
            departs.append(value)
        return departs   

    def set_department(self,department):
        '''
        Takes a department as a string and passes corresponding department code to form
        '''
        dept = [k for k,v in self.Departments.items() if v.casefold() == department.casefold()]
        self.page.set("ctl00$ContentPlaceHolder1$FormView1$DropDownList_Department", dept[0])
        
    def get_years(self):
        '''
        Compares current date to end of May Term. 
        If date is before june 1st, it will return an array of the past 2 years and future year
        if date is past, it will return last year and future 2 years. (Used for Determining Valid Years for Terms)
        '''
        dt = datetime.now()
        current_year = int('{:%Y}'.format(dt))
        end_of_term = datetime(current_year, 6, 1)
        if dt < end_of_term:
            return [current_year-2, current_year-1, current_year, current_year+1]
        else:
            return [current_year-1, current_year, current_year+1, current_year+2]
    
    def get_terms(self):    
        '''
        Returns array of current terms.
        '''
        return self.Terms

    def set_term(self,year,semester):
        '''
        Takes year (int) and semester (str)
        passes parameters to form as concatencated str
        '''
        term = str(year)+' '+semester
        self.page.set("ctl00$ContentPlaceHolder1$FormView1$DropDownList_Term", term)

    def get_times(self):
        '''
        Returns array of available class times
        '''
        times = []
        for key,value in self.Times.items():
                times.append(value)
        return times 

    def set_time(self,period):
        '''
        Takes time as parameter and passes the proper form value to search form
        '''
        for k,v in self.Times.items():
            if period in v:
                self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_MeetingTime', k)

    def get_ed(self):
        '''
        Returns array of possible essential ed requirements
        '''
        eds = []
        for key,value in self.ED.items():
                eds.append(value)
        return eds

    def set_ed(self,ed):
        '''
        Takes 'ed' as string and passes to form for Essential Ed.
        '''
        for k,v in self.Times.items():
            if ed in v:
                self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_EssentialEd', k)
    
    def get_cd(self):
        '''
        Returns list of Cultural-Diversity Requirment Options
        '''
        cd = []
        for key,value in self.CD.items():
            cd.append(value)
        return cd

    def set_cd(self,option):
        '''
        Takes option as a str, evaluates that self.CD contains the parameter
        If so, passes it to form to set Culural Diversity field.
        '''
        cdo = [k for k,v in self.CD.items() if v.casefold() == option.casefold()]
        self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_CulturalDiversity', cdo[0])

    def set_wi(self, wi):
        '''
        Takes 'wi' as bool and sets fields with boolean.
        '''
        if wi == True:
            self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_WritingIntensive', 'WI')
        else:
            self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_WritingIntensive', 'none')

    def is_pf(self,pf):
        '''
        Takes parameter 'pf' as bool
        passes corresponding T/F to Pass/Fail Field
        '''
        if pf == True:
            self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_PassFail', 'PF')
        elif pf == False:
            self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_PassFail','none')

    def get_instructors(self):
        '''
        Determines instructors from Search form HTML,
        Creates a dictionary based on their ID:Names
        Returns an array all names.
        '''
        for item in self.option_list:
            value = str(item).split('"') #used " as delimeter due to , and ' being used in the course data
            if value[1].isdigit():
                self.Instructor[value[1]] = item.get_text()
        proflist = []
        for key,value in self.Instructor.items():
            proflist.append(value)
        return proflist

    def set_instructor(self, name):
        '''
        Takes parameter name as a str, cross-references instructor dictionary
        passes Instructor ID to instructor field.
        '''
        #transform name to id_num
        for k,v in self.Instructor.items():
            if name in v:
                self.page.set('ctl00$ContentPlaceHolder1$FormView1$DropDownList_Instructor', k)

    def couse_open(self, option):
        '''
        Takes parameter option as bool, passes bool to field
        '''
        self.page.set('ctl00$ContentPlaceHolder1$FormView1$CheckBox_OpenCourses', option)
    
    def search_form(self):
        '''
        Identifies form submission button, submits form and
        extracts the courses found in the results.
        Returns the courses in the form of a Tuple containing Tuples.
        '''
        self.course_list = []
        course = []
        counter = 0
        self.page.choose_submit('ctl00$ContentPlaceHolder1$FormView1$Button_FindNow')
        self.browser.submit_selected()
        table_of_data = self.browser.get_current_page().find('table')
        data = table_of_data.get_text('td').replace('\n', '').replace('td', '|')
        data2 = data.split('|')
        for index in data2[18:]:
            if counter < 15:
                if index == '':
                    continue
                else:
                    if index == '\xa0' or index == ' ':
                        index = 'Null'
                    course.append(index)
                    counter +=1
            else:
                if index == '\xa0' or index == ' ':
                    index = 'Null'
                course.append(index)
                self.course_list.append(tuple(course))
                course = []
                counter = 0
        for item in self.course_list:
           self.data.append([item[5],item[6],item[7],item[8],item[9],item[10],item[11],item[12],item[14]])
        
        return self.data

    def display_browser(self):
        '''
        Displays cached copy of webpage
        '''
        self.browser.launch_browser()
    
    def save_file(self):
        '''
        Saves the results of the search to a .csv file
        '''
        
        with open('test_data.txt', 'w') as handler: #output of scraped data
            print("writing file...")
            
            for course in self.course_list:
                handler.write('%s\n' % list(course))           
